### PR TITLE
Preserve pinned executables

### PR DESCRIPTION
**Pull request by thoemmi**
If the user pinned the app to the taskbar, the upgrade process should update the shortcuts.

The first two commits 035a311 and d16793c delete the old pinned shortcuts and create new ones afterwards.

This however changed the order of items in the taskbar, so I changed it in the third commit (95f864e). Now the code just fixes the existing shortcuts (and deletes the obsolete ones).

Finally 54d3246 provides a unit test.

Either accept this PR or modify the code according to your needs.

### CODE REVIEW

File: src/Shimmer.Tests/Client/ApplyReleasesTests.cs [hunk 1]
```diff
@@ -249,6 +249,52 @@ public void IfAppSetupThrowsWeFailTheInstall()
...
+                updateInfo = UpdateInfo.Create(baseEntry, new[] { latestFullEntry }, "dontcare", FrameworkVersion.Net40);
+                using (fixture.AcquireUpdateLock()) {
+                    fixture.ApplyReleases(updateInfo).ToList().First();
+                }
+
+                var newExecutable = Path.Combine(tempDir, "theApp", "app-1.1.0.0", "SampleUpdatingApp.exe");
+                File.Exists(newExecutable).ShouldBeTrue();
+                TaskbarHelper.IsPinnedToTaskbar(newExecutable).ShouldBeTrue();
+
+                var retries = 20;
```
Comment (anaisbetts) [inline src/Shimmer.Tests/Client/ApplyReleasesTests.cs]: Can you use the Retry() extension method here instead?

File: src/Shimmer.Client/TaskbarHelper.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Shimmer.Client {
+    public static class TaskbarHelper {
+        public static bool IsPinnedToTaskbar(string executablePath) {
+            var taskbarPath = Path.Combine(
+                Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
+                "Microsoft\\Internet Explorer\\Quick Launch\\User Pinned\\TaskBar");
```
Comment (anaisbetts) [inline src/Shimmer.Client/TaskbarHelper.cs:11]: This is Danger Zone™, we're hacking into implementation details of Win7. However, I don't think there's any proper way to do this - might be best to gate this behind a version check (i.e. only do this on Win7 + Win8)

**Comment by anaisbetts** [comment]
This is a great PR!

File: src/Shimmer.Client/TaskbarHelper.cs (hunk 2 above)
Comment (thoemmi) [inline src/Shimmer.Client/TaskbarHelper.cs:11]: Sounds reasonable, I'll add a check.

File: src/Shimmer.Tests/Client/ApplyReleasesTests.cs (hunk 1 above)
Comment (thoemmi) [inline src/Shimmer.Tests/Client/ApplyReleasesTests.cs]: I'll had a short glance at `Utility.Retry`, but it relies on a thrown exception, not a boolean condition. May I add such an implementation?

Comment (anaisbetts) [inline src/Shimmer.Tests/Client/ApplyReleasesTests.cs]: Can you make pinUnpin throw if it doesn't work?

**Comment by anaisbetts** [comment]
One more question, in what way does the NuPkg fixture change?

**Comment by thoemmi** [comment]
_SampleUpdatingApp.1.1.0.0.nupkg_ is just a copy (and rename) from _src\SampleUpdatingApp\SampleReleasesFolder_. To test the upgrade of pinned shortcuts, I first needed a 1.0 version to create a shortcut with, and a version 1.1, so that the shortcut must to link to the new location after the upgrade.

**Comment by anaisbetts** [comment]
:cool:, thanks for the PR!

**Pull request merged**